Rework/redesign/reimplement mutation debugging/tracking for simulation

alexmiller-apple opened the pull request:
As a relatively unknown debugging tool for simulation tests, one could
have simulation print when a particular key is handled in various stages
of the commit process.  This functionality was enabled by changing a 0
to a 1 in an #if, and changing a constant to the key in question.

This is intended to speed up a common path of debugging simulation tests,
where one key that has a wrong value is found, and then traced through FDB.
For example, if CycleTest failed because key `foo` has an incorrect value, why
is it wrong?  Where did the value come from?  What happened to that key
during the test?

As a proxy and storage server handle mutations, they call debugMutation
or debugKeyRange, which then checks against the mutation against the key
in question, and logs if they match.  A mixture of printfs and
TraceEvents would then be emitted, and for this to actually be usable,
one also needs to comment out some particularly spammy debugKeyRange()
calls.  This results in some difficult to use and incomplete tracking.

This PR reworks the API of debugMutation/debugKeyRange, pulls it out
into its own file, and trims what is logged by default into something
useful and understandable:
* debugMutation() now returns a TraceEvent, that one can add more details to before it is logged.
* Data distribution and storage server cleanup operations are no longer logged by default

So the API is cleaner, and the output is both consistent and clearer.

This also adds a new debugMessagesAndTags, and track mutations in more places.

Like:
* Leaving the proxy
* Entering the TLog
* Leaving the TLog
* Being read on a cursor

All of this brought to you by TagsAndMessage!

This also slides in a minor optimization as to how mutations are serialized per target log, because it was convenient.

jzhou77 commented on fdbserver/TLogServer.actor.cpp:
```diff
@@ -1335,7 +1337,12 @@ void peekMessagesFromMemory( Reference<LogData> self, TLogPeekRequest const& req
 		}
 
+		// We need the 4 byte length prefix to be a TagsAndMessage format, but that prefix is added as part of StringRef serialization.
+	  int offset = messages.getLength();
```
The leading space is not aligned.

jzhou77 commented on fdbserver/MutationTracking.cpp:
```diff
@@ -0,0 +1,100 @@
+#error "You cannot use mutation tracking in a clean/release build."
+#endif
+
+StringRef debugKey = LiteralStringRef( "\xff/globals/lastEpochEnd" );
```
This was `LiteralStringRef( "" );` in the original code. Maybe we should keep using it?

And it would be nice to add some comment for these two keys so that if a developer need to look into specific version/key/range, he/she knows how to use these two keys.

jzhou77 commented on fdbserver/MutationTracking.h:
```diff
@@ -0,0 +1,34 @@
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/CommitTransaction.h"
+
+#define MUTATION_TRACKING_ENABLED 0
```
Maybe moving this to the .cpp file is better, because then all modifications are just in one file and only one file needs to be recompiled.

satherton replied:
In my proposed solutions for getting rid of the overhead from constructing a trace event and making `detail()` calls on it when mutation tracking is disabled, this macro would have to stay in the header.

dongxinEric commented on fdbserver/MutationTracking.cpp:
```diff
@@ -0,0 +1,100 @@
+TraceEvent debugMutationEnabled( const char* context, Version version, MutationRef const& mutation ) {
+	if ((mutation.type == mutation.ClearRange || mutation.type == mutation.DebugKeyRange) &&
+			((mutation.param1<=debugKey && mutation.param2>debugKey) || (mutation.param1<=debugKey2 && mutation.param2>debugKey2))) {
+		return std::move(TraceEvent("MutationTracking").detail("At", context).detail("Version", version).detail("MutationType", typeString[mutation.type]).detail("KeyBegin", mutation.param1).detail("KeyEnd", mutation.param2));
```
Since it's debug code I feel it's safe to assume the complier supports RVO.

satherton commented on fdbserver/MutationTracking.cpp:
```diff
@@ -0,0 +1,100 @@
+	return debugMessagesAndTagsEnabled( context, version, commitBlob );
+}
+#else
+TraceEvent debugMutation( const char* context, Version version, MutationRef const& mutation ) { return std::move(TraceEvent()); }
```
Why the explicit std::move, is there doubt that RVO will occur?

dongxinEric replied:
It's at the mercy of compliers.

satherton replied:
I could be wrong but I thought RVO would handle things like `TraceEvent e = debugkeyRange(..)` in this scenario.

ajbeamon replied:
If it doesn't do the optimization, are you going to get an extra trace event logged due to the destruction?

dongxinEric replied:
I think so.

satherton replied:
@dongxinEric Sure, but if we are assuming the worst then I guess a lot of other code should probably be more explicit about such things as well.

In any case, I'm not sure that this pattern is what we want here.  Moving a TraceEvent sounds better than copying one, but even constructing one is not exactly trivial and moving one [isn't either](https://github.com/apple/foundationdb/blob/master/flow/Trace.cpp#L764).

Since disabled TraceEvents do not log details, perhaps when mutation tracking is disabled these methods could return a reference to a static TraceEvent that exists to absorb/ignore detail calls, rather than initializing a new one only to have some ignored details added before it is destroyed.  

Or perhaps even better would be for these methods to return a type that looks like a TraceEvent but does nothing, so even the detail() argument initialization should be optimized out.  If the callers used `auto e = fun(...)` rather than `TraceEvent e = fun(...)` this would work, though it is at least a bit awkward.

However, these suggestions still have an overhead problem where statements like `e.detail("Foo", x.toString())` (which I see in this changeset) will still execute the `.toString()` call (unless compiler and link time optimization is being amazing) even though the detail won't be added.

I should point out that my assumption here is that since every key that passes through every role of fdb is going to have `debugMutation()` called on it several times, constructing a default TraceEvent and making some `detail()` calls each time that happens is not a trivial amount of overhead.

satherton replied:
Since the return type of the method is `TraceEvent` and not `TraceEvent &&`, I'm not even sure the `std::move()` is doing anything.

Daniel-B-Smith replied:
As of C++17, the RVO in this case is mandatory: https://en.cppreference.com/w/cpp/language/copy_elision

We could even go so far as deleting the copy/move constructors, and these functions would still work.

alexmiller-apple replied:
> Why the explicit std::move, is there doubt that RVO will occur?

Because if it's not there, then gcc threw an error that the copy constructor was private/deleted.

alexmiller-apple replied:
If we really want the overhead of this to go away, I can instead

```
#define UNHYGENIC_DEBUG_MUTATION(context, version, mutationref) !MUTATION_TRACKING_ENABLED ? (void)TraceEvent() : debugMutationEnabled(context, version, mutationref)

UNHYGENIC_DEBUG_MUTATION("ProxyCommit", version, m).detail("Foo", foo);
```

But I was hoping it wouldn't come to that.  I'll see if I can play more with c++17 and RVO to see if I can massage the code into something that gcc will accept and doesn't require a move or copy.

My overall preference is to keep the code as something that's compiled and then dead code eliminated, so that people are forced to maintain it as .e.g variable names change.

satherton replied:
I do think we need to eliminate this overhead, I would be very surprised if this much work, many times over, per KV pair, is trivial.

I absolutely agree with the goal of having code that is always compiled but that is optimized out when not needed.

My understanding is that the compiler would have no way of knowing that the `.detail()` calls on the default-constructed TraceEvent have no side effects and can safely be removed, since the `detail()` methods are not defined in `Trace.h`, so I don't think it can just decide to not call them.

There is of course always the option of a run-time enabled check using `if` rather than `#if`, such as
```
if(MUTATION_TRACKING_ENABLED) {
...
}
```
Since `MUTATION_TRACKING_ENABLED` is a constant the code block should be compiled but optimized out.

Here's another option that looks cleaner and I think would work.  First, MUTATION_TRACKING_ENABLED has to be defined to 0 if it is not already defined.  Then make `TraceEvent` cast to `bool` defined as
```
operator bool() const {
  return enabled;
}
```
which I think makes sense in a bool context.  Then, we can do this:
```
MUTATION_TRACKING_ENABLED && debugMutation(...).detail(...).detail(..);
```
and the right hand side should be compiled and then optimized out.

alexmiller-apple replied:
I ended up doing a hybrid of the two.  I added a macro that does `MUTATION_TRACKING_ENABLED && debugMutation(...)`, so writing `DEBUG_MUTATION(x,y,z).detail("foo", "bar")` expands to `ENABLED && debugMutation(x,y,z).debug("foo", "bar")`.  This breaks being able to do `TraceEvent&& e = DEBUG_MUTATION(x,y,z)`, so one has to manually `if (MUTATION_TRACKING_ENABLED)`, and then call `debugMutation` instead.  One shouldn't need to do this often, so I think it's okay.

xumengpanda commented on fdbserver/RestoreLoader.actor.cpp:
```diff
@@ -463,25 +464,25 @@ ACTOR Future<Void> sendMutationsToApplier(VersionedMutationsMap* pkvOps, int bat
-						    .detail(format("UID%d", i).c_str(), uid.toString());
-						i++;
+				{
+					TraceEvent&& e = debugMutation("RestoreLoaderDebugSplit", commitVersion.version, kvm);
```
Could you change the event name to `RestoreLoaderDebugRangeToAppliers`, which better reflects the event details.

xumengpanda commented on fdbserver/MutationTracking.cpp:
```diff
@@ -0,0 +1,100 @@
+			br >> m;
+			TraceEvent&& event = debugMutation(context, version, m);
+			if (event.isEnabled()) {
+				return std::move(event.detail("MessageTags", msg.tags));
```
Is it expected that we only return the first matched mutation in a message?
What if there are more than one mutation valid in the debug key range? Not outputting the existed-but-missed mutation could lead to confusion, IMO.

alexmiller-apple replied:
I think the right thing to do here longer term would be to make a TraceEventGroup that has the same API as a trace event, and fowards all calls to the N trace events that it internally contains.  That would have to be later work though.  I'll at least document this for now.

xumengpanda commented on fdbserver/storageserver.actor.cpp:
```diff
@@ -3050,7 +3043,8 @@ bool StorageServerDisk::makeVersionMutationsDurable( Version& prevStorageVersion
 		VersionUpdateRef const& v = u->second;
 		ASSERT( v.version > prevStorageVersion && v.version <= newStorageVersion );
-		debugKeyRange("makeVersionMutationsDurable", v.version, allKeys);
+		// TODO(alexmiller): Update to version tracking.
```
Could you explain a bit what `version tracking` is when you are updating to it?

alexmiller-apple commented:
I think I have addressed most of the feedback here.  If someone works out a better debugMutation API in the future, I'm all for it, but this seems like a large enough step forward and a contained amount of work that it's better to land this now, and folk are welcome to continue to work on improvements later.
